internal/query/roles_v2.go

```diff
@@ -0,0 +1,668 @@
+
+		logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+
+		// At this point, spicedb changes have already been applied.
+		// Attempting to rollback could result in failures that could result in the same situation.
+		//
+		// TODO: add spicedb rollback logic along with rollback failure scenarios.
+
+		return types.Role{}, err
+	}
+
+	role.Name = dbRole.Name
+	role.CreatedBy = dbRole.CreatedBy
+	role.UpdatedBy = dbRole.UpdatedBy
+	role.ResourceID = dbRole.ResourceID
+	role.CreatedAt = dbRole.CreatedAt
+	role.UpdatedAt = dbRole.UpdatedAt
+	role.Actions = newActions
+
+	return role, nil
+}
+
+func (e *engine) DeleteRoleV2(ctx context.Context, roleResource types.Resource) error {
+	ctx, span := e.tracer.Start(ctx, "engine.DeleteRoleV2")
+	defer span.End()
+
+	dbRole, err := e.store.GetRoleByID(ctx, roleResource.ID)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		return err
+	}
+
+	roleOwner, err := e.NewResourceFromID(dbRole.ResourceID)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		return err
+	}
+
+	dbCtx, err := e.store.BeginContext(ctx)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		return err
+	}
+
+	err = e.store.LockRoleForUpdate(dbCtx, roleResource.ID)
+	if err != nil {
+		sErr := fmt.Errorf("failed to lock role: %s: %w", roleResource.ID, err)
+
+		span.RecordError(sErr)
+		span.SetStatus(codes.Error, sErr.Error())
+
+		logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+
+		return err
+	}
+
+	// 1. delete role from permission-api DB
+	if _, err = e.store.DeleteRole(dbCtx, roleResource.ID); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+
+		return err
+	}
+
+	// 2. delete role relationships from spice db
+	const deleteErrsBufferSize = 2
+
+	wg := &sync.WaitGroup{}
+	errs := make(chan error, deleteErrsBufferSize)
+
+	// 2.a remove all relationships from this role
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+
+		delRoleRelationshipReq := &pb.DeleteRelationshipsRequest{
+			RelationshipFilter: &pb.RelationshipFilter{
+				ResourceType:       e.namespaced(e.rbac.RoleResource.Name),
+				OptionalResourceId: roleResource.ID.String(),
+			},
+		}
+
+		if _, err := e.client.DeleteRelationships(ctx, delRoleRelationshipReq); err != nil {
+			errs <- err
+		}
+	}()
+
+	// 2.b remove all relationships to this role from its owner
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+
+		ownerRelReq := &pb.DeleteRelationshipsRequest{
+			RelationshipFilter: &pb.RelationshipFilter{
+				ResourceType: e.namespaced(roleOwner.Type),
+				OptionalSubjectFilter: &pb.SubjectFilter{
+					SubjectType:       e.namespaced(e.rbac.RoleResource.Name),
+					OptionalSubjectId: roleResource.ID.String(),
+				},
+			},
+		}
+
+		if _, err := e.client.DeleteRelationships(ctx, ownerRelReq); err != nil {
+			errs <- err
+		}
+	}()
```

Comment:
As before, I'm not really sold on the concurrency here given the added complexity.